prog/test/github_runner.rb

```diff
@@ -0,0 +1,150 @@
+    hop_download_boot_images
+  end
+
+  label def download_boot_images
+    frame["test_cases"].each do |test_case|
+      bud Prog::DownloadBootImage, {"subject_id" => vm_host_id, "image_name" => tests[test_case]["image_name"]}
+    end
+
+    hop_wait_download_boot_images
+  end
+
+  label def wait_download_boot_images
+    reap
+    hop_trigger_test_runs if leaf?
+    donate
+  end
+
+  label def trigger_test_runs
+    test_runs.each do |test_run|
+      trigger_test_run(test_run["repo_name"], test_run["workflow_name"], test_run["branch_name"])
+    end
+
+    # To make sure that test runs are triggered
+    sleep 30
+
+    hop_check_test_runs
+  end
+
+  label def check_test_runs
+    test_runs.each do |test_run|
+      conclusion = test_run_conclusion(test_run["repo_name"], test_run["workflow_name"], test_run["branch_name"])
+      if FAIL_CONCLUSIONS.include?(conclusion)
+        fail_test "Test run for #{test_run["repo_name"]}, #{test_run["workflow_name"]}, #{test_run["branch_name"]} failed with conclusion #{conclusion}"
+      elsif IN_PROGRESS_CONCLUSIONS.include?(conclusion) || conclusion.nil?
+        nap 15
+      end
+    end
+
+    hop_finish
+  end
+
+  label def finish
+    cleanup
+    pop "GithubRunnerGroup tests are finished!"
+  end
+
+  label def failed
+    cleanup
+    nap 15
+  end
+
+  def trigger_test_run(repo_name, workflow_name, branch_name)
+    url = "repos/#{repo_name}/actions/workflows/#{workflow_name}/dispatches"
+    payload = {ref: branch_name}
+
+    begin
+      client.post(url, payload)
+    rescue Octokit::Error => e
+      fail_test "Can not trigger workflow for #{repo_name}, #{workflow_name}, #{branch_name} due to #{e.message}"
+    end
+  end
+
+  def test_run_conclusion(repo_name, workflow_name, branch_name)
+    run_id = get_latest_run_id(repo_name, workflow_name, branch_name)
+    get_run_conclusion(repo_name, run_id)
+  end
+
+  def get_latest_run_id(repo_name, workflow_name, branch)
+    runs = client.workflow_runs(repo_name, workflow_name, {branch: branch})
+    latest_run = runs[:workflow_runs].first
+    latest_run[:id]
+  end
+
+  def get_run_conclusion(repo_name, run_id)
+    run = client.workflow_run(repo_name, run_id)
+    run[:conclusion]
+  end
+
+  def cleanup
+    Project[frame["github_service_project_id"]]&.destroy
+    Project[frame["github_test_project_id"]]&.destroy
+    cancel_test_runs
+  end
+
+  def cancel_test_runs
+    test_runs.each do |test_run|
+      cancel_test_run(test_run["repo_name"], test_run["workflow_name"], test_run["branch_name"])
+    end
+  end
+
+  def cancel_test_run(repo_name, workflow_name, branch_name)
+    run_id = get_latest_run_id(repo_name, workflow_name, branch_name)
+    begin
+      client.cancel_workflow_run(repo_name, run_id)
+    rescue
+      puts "Workflow run #{run_id} for #{repo_name} has already been finished"
+    end
+  end
+
+  def tests
+    @@tests ||= YAML.load_file("config/github_runner_e2e_tests.yml").to_h { [_1["name"], _1] }
```

Comment:
I still have doubts about putting test cases in the YAML instead of a constant variable. It seems too fragmented to me.

Reply:
I though about it and I slightly prefer having it in config file instead of having it defined on the code itself as I find understanding and updating config easier. Though, if you have strong opinion I'm ok to change it as well.